File: CSES/Problem Set/Mathematics/Inversion Probability.py
from decimal import Decimal, getcontext
getcontext().prec=50
res=Decimal(0)
con=[Decimal(0)]*100
n=int(input())
X=list(map(Decimal,input().split()))
for x in X:
    for j in range(100):res+=con[j]*min(j,x)/x
    for j in range(int(x)):con[j]+=1/x
print(res.quantize(Decimal("0.000001")))

File: CSES/Problem Set/Mathematics/Inversion Probability.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef __float128 ld;
int n,x;
ld con[100],res;
int main(){
    ios_base::sync_with_stdio(0);
    cin.tie(0);cout.tie(0);
    cin >> n;
    for(int i=0;i<n;i++){
        cin >> x;
        for(int k=0;k<100;k++)res+=con[k]*min(k,x)/x;
        for(int j=0;j<x;j++)con[j]+=(ld)1/x;
    }
    cout << fixed << setprecision(6) << (long double)res;
}